### PR TITLE
Adds support for generating projects that use AndroidX support libraries

### PR DESCRIPTION
**Pull request by josh-burton**
## Description

This PR modifies the Flutter tool with support for generating projects that use AndroidX support libraries rather than the legacy support library.

This is a source of pain in Flutter development and now that AndroidX is 1.0 all new Flutter projects should be using it.

Projects generated with AndroidX libraries also use Jetifier which allows any dependencies (e.g. plugins) that still use the legacy support libraries to be upgrade to use AndroidX.

With this change, the flutter tool has the following changes:

* Adds an `--androidx` flag when creating flutter projects. This is disabled by default, and can be enabled via `--androidx`
* Adds a new `androidX` property under the module section of a pubspec.yaml file. This is to control whether the generated module project uses AndroidX.
e.g.
```
flutter:
  uses-material-design: true

  module:
    androidX: true
    androidPackage:
    iosBundleIdentifier:
```

## Related Issues

* https://github.com/flutter/flutter/issues/28805

## Tests

* create_test.dart/androidx app project
* create_test.dart/non androidx app project
* create_test.dart/androidx plugin project
* create_test.dart/non androidx plugin project

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by jiechic** [comment]
Why haven't merged yet?
Is it necessary use support library by default and Optional androidx support before code merge?

**Comment by athornz** [comment]
@dnfield I've seen you comment about AndroidX support a few other places. Is it possible this PR may be merged or is there other internal work happening around it?

**Comment by jiechic** [comment]
Why should this PR not be merged?

**Comment by athornz** [comment]
@matthew-carroll would you be the right person to review this?

**Comment by matthew-carroll** [comment]
@xster you're probably in the best position to make a call on this particular solution.

**Comment by jiechic** [comment]
#28805

**Comment by xster** [comment]
@dnfield @mklim, could you guys take a pass at this?

I'm out for a week. I'll check this wrt add-to-app when I'm back.

**Comment by dnfield** [comment]
Thanks for the contribution!

We've had a number of issues with jetifier not working with our current gradle setup.  I'm a bit curious about how this impacts that - are you able to consume a plugin that uses the support libraries with this project structure successfully?  

My main concern here is that it's not clear what benefit this really offers to someone developing a Flutter app at this point - unless it's helping with that issue.

**Comment by athornz** [comment]
Hey Dan,

With this setup, a project will work with both AndroidX and Support dependencies.

Without this, I believe Flutter projects that use plugins with AndroidX dependencies need to be upgraded to AndroidX.

I'm using this branch in a Flutter project using Jetifier and all is working well.

So the benefit is that new Flutter projects using AndroidX should just work with any plugin. Do you have any more information about the issues with Jetifier that you've had? I'd like to test this branch to make sure it solves those issues.

**Comment by jiechic** [comment]
@athornz 
I think it should use support library as the default then androidX library is optional for old support library flutter module project update

**Comment by athornz** [comment]
Since support library is deprecated I think AndroidX should be the default. If a Flutter project with AndroidX works with Jetifier then there is no reason to use the support libraries.

We should be making use of Jetifier as Google implemented as a good way to update to AndroidX and be compatible with support libraries.

**Comment by mklim** [comment]
Thanks for the contribution! @athornz I agree, AndroidX with Jetifier should ideally be our default to get rid of any support issues. The problem we've run into so far is that Jetifier doesn't work with how we build Flutter apps and include plugins: flutter/flutter#32714. You can see this by making a new template with this PR or by migrating an app manually, then depending on a an older plugin (like `firebase_core: 0.1.0`), then running `flutter build apk`. 

Even with Jetifier being broken for us @dnfield I think this is probably an improvement. At least all of `flutter/plugins` are migrated to AndroidX at this point, so with that it makes sense to me to have the default template be migrated too. Right now people still need to migrate manually.

**Review by tvolkert — COMMENTED** [review]
Thanks for the contribution @athornz - this is great!

@dnfield having just spent 5 hours wrestling with AndroidX issues in my project, I think this change will help enable new projects to be built.

File: packages/flutter_tools/templates/app/android.tmpl/gradle.properties
Comment (dnfield) [inline packages/flutter_tools/templates/app/android.tmpl/gradle.properties]: Why is this getting deleted?  I believe this was put in place to fix a frequent issue where gradle would crash due to lack of memory.

File: packages/flutter_tools/templates/module/android/gradle/gradle.properties.copy.tmpl
Comment (dnfield) [inline packages/flutter_tools/templates/module/android/gradle/gradle.properties.copy.tmpl]: Ditto

File: packages/flutter_tools/templates/plugin/android.tmpl/gradle.properties
Comment (dnfield) [inline packages/flutter_tools/templates/plugin/android.tmpl/gradle.properties]: Ditto

File: packages/flutter_tools/templates/app/android.tmpl/gradle.properties
Comment (tvolkert) [reply packages/flutter_tools/templates/app/android.tmpl/gradle.properties]: It's not being deleted - it moved from a static file to a template.

File: packages/flutter_tools/templates/module/android/gradle/gradle.properties.copy.tmpl
Comment (tvolkert) [reply packages/flutter_tools/templates/module/android/gradle/gradle.properties.copy.tmpl]: Ditto 🙂

File: packages/flutter_tools/templates/plugin/android.tmpl/gradle.properties
Comment (tvolkert) [reply packages/flutter_tools/templates/plugin/android.tmpl/gradle.properties]: Ditto 🙂

File: packages/flutter_tools/templates/module/android/library/Flutter.tmpl/src/main/java/io/flutter/facade/Flutter.java.tmpl
Comment (tvolkert) [reply packages/flutter_tools/templates/module/android/library/Flutter.tmpl/src/main/java/io/flutter/facade/Flutter.java.tmpl]: They're not removed - they just got re-ordered given the new conditional blocks.

Comment (dnfield) [reply packages/flutter_tools/templates/module/android/library/Flutter.tmpl/src/main/java/io/flutter/facade/Flutter.java.tmpl]: Ahh I missed where they landed.

File: packages/flutter_tools/templates/app/android.tmpl/gradle.properties
Comment (dnfield) [reply packages/flutter_tools/templates/app/android.tmpl/gradle.properties]: I see it now, sorry about that.

File: packages/flutter_tools/lib/src/commands/create.dart
Comment (athornz) [reply packages/flutter_tools/lib/src/commands/create.dart]: I've updated the PR description.

File: packages/flutter_tools/lib/src/project.dart
Comment (athornz) [reply packages/flutter_tools/lib/src/project.dart]: Doesn't look like it, I've removed it.

File: packages/flutter_tools/lib/src/flutter_manifest.dart
Comment (athornz) [reply packages/flutter_tools/lib/src/flutter_manifest.dart]: ✔️

File: packages/flutter_tools/lib/src/project.dart
Comment (athornz) [reply packages/flutter_tools/lib/src/project.dart]: ✔️

File: packages/flutter_tools/lib/src/project.dart
Comment (athornz) [reply packages/flutter_tools/lib/src/project.dart]: ✔️

File: packages/flutter_tools/templates/app/android.tmpl/gradle.properties.tmpl
Comment (athornz) [reply packages/flutter_tools/templates/app/android.tmpl/gradle.properties.tmpl]: ✔️

File: packages/flutter_tools/templates/module/android/gradle/gradle.properties.tmpl
Comment (athornz) [reply packages/flutter_tools/templates/module/android/gradle/gradle.properties.tmpl]: ✔️

File: packages/flutter_tools/templates/plugin/android.tmpl/gradle.properties.tmpl
Comment (athornz) [reply packages/flutter_tools/templates/plugin/android.tmpl/gradle.properties.tmpl]: ✔️

**Comment by athornz** [comment]
Thanks @dnfield and @tvolkert for the review - I think I've addressed all the comments now but let me know if there is anything else.

File: packages/flutter_tools/lib/src/project.dart
Comment (tvolkert) [reply packages/flutter_tools/lib/src/project.dart]: Now this getter can just return `parent.usesAndroidX`

**Review by tvolkert — APPROVED** [review]
LGTM modulo one small nit.

File: packages/flutter_tools/lib/src/project.dart
Comment (athornz) [reply packages/flutter_tools/lib/src/project.dart]: good point, done!

**Comment by tvolkert** [comment]
LGTM.  Will land on green.

Thanks again!

**Comment by athornz** [comment]
No problem, thank you!

**Review by dnfield — APPROVED** [review]

**Comment by jonahwilliams** [comment]
Looks like you're synced to a bad commit of master, you'll need to fetch and update

**Comment by athornz** [comment]
@jonahwilliams I've rebased on master now.

**Comment by mcxinyu** [comment]
Flutter.jar is not compatible with androidX

**Comment by truongsinh** [comment]
@mcxinyu 
> Flutter.jar is not compatible with androidX

Currently I have a the app crashed:
```
    java.lang.NoClassDefFoundError: Failed resolution of: Landroidx/lifecycle/DefaultLifecycleObserver;
        at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:149)
```

is it related to flutter.jar? cc @athornz @dnfield @tvolkert

**Comment by blasten** [comment]
That is because we are missing AndroidX transitive dependencies from the engine. We have a fix in progress. cc @matthew-carroll

**Comment by blasten** [comment]
In the meanwhile, you can the dependencies manually to `android/build.gradle` as indicated on https://developer.android.com/jetpack/androidx/releases/lifecycle#declaring_dependencies

**Comment by truongsinh** [comment]
The crash still happens even after I added the following snippet to `app/build.gradle`

```groovy

    // known issue https://github.com/flutter/flutter/pull/31028#issuecomment-516902006
    def lifecycle_version = "2.0.0"
    // ViewModel and LiveData
    implementation "androidx.lifecycle:lifecycle-extensions:$lifecycle_version"
    // alternatively - just ViewModel
    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:$lifecycle_version"
    // alternatively - just LiveData
    implementation "androidx.lifecycle:lifecycle-livedata:$lifecycle_version"
    implementation "androidx.lifecycle:lifecycle-runtime:$lifecycle_version"
    kapt "androidx.lifecycle:lifecycle-compiler:$lifecycle_version"
    // optional - ReactiveStreams support for LiveData
    implementation "androidx.lifecycle:lifecycle-reactivestreams-ktx:$lifecycle_version"
    // optional - Test helpers for LiveData
    testImplementation "androidx.arch.core:core-testing:$lifecycle_version"
```

**Comment by blasten** [comment]
@truongsinh would you mind creating an issue with a repro test case? Thanks!

**Comment by truongsinh** [comment]
> @truongsinh would you mind creating an issue with a repro test case? Thanks!

I might take a while, but sure.

**Comment by truongsinh** [comment]
Actually, trying to reproduce the error on a sample, I encoutered a different problem (here's the code https://github.com/truongsinh/android_flutter_host/tree/built-in-androidx)

```
Cannot open a library at 'FileMapping(from=/Users/truongsinh/Dev/flutter/android_flutter_host/flutter_embedding/.android/Flutter/build/intermediates/flutter/debug/libs.jar, to=/Users/truongsinh/.gradle/caches/transforms-1/files-1.1/libs.jar/9a5b6ea5d0ab9a291353d9eb6a6ec26b/jetified-libs.jar)'
```

**Comment by truongsinh** [comment]
@athornz @blasten any update?

**Comment by truongsinh** [comment]
update from my side,

```
Cannot open a library at 'FileMapping(from=/Users/truongsinh/Dev/flutter/android_flutter_host/flutter_embedding/.android/Flutter/build/intermediates/flutter/debug/libs.jar, to=/Users/truongsinh/.gradle/caches/transforms-1/files-1.1/libs.jar/9a5b6ea5d0ab9a291353d9eb6a6ec26b/jetified-libs.jar)'
```

is solved **by work around**, I have to `cd flutter_embedding/.android && ./gradlew assembleDebug` to generate those jars file, and then I can reproduce 

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: pro.truongsinh.flutter.android_flutter_host.free, PID: 11100
    java.lang.NoClassDefFoundError: Failed resolution of: Landroidx/lifecycle/DefaultLifecycleObserver;
        at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:149)
        at pro.truongsinh.flutter.android_flutter_host.FlutterEmbeddingActivity$Companion.init(FlutterEmbeddingActivity.kt:168)
        at pro.truongsinh.flutter.android_flutter_host.MainActivity.onCreate(MainActivity.kt:17)
        at android.app.Activity.performCreate(Activity.java:7802)
        at android.app.Activity.performCreate(Activity.java:7791)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1299)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3243)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3407)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7343)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:933)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "androidx.lifecycle.DefaultLifecycleObserver" on path: DexPathList[[zip file "/data/app/pro.truongsinh.flutter.android_flutter_host.free-R3e6unzzN8OPsBvhFb5V2w==/base.apk"],nativeLibraryDirectories=[/data/app/pro.truongsinh.flutter.android_flutter_host.free-R3e6unzzN8OPsBvhFb5V2w==/lib/arm64, /data/app/pro.truongsinh.flutter.android_flutter_host.free-R3e6unzzN8OPsBvhFb5V2w==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:196)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at io.flutter.embedding.engine.FlutterEngine.<init>(FlutterEngine.java:149) 
        at pro.truongsinh.flutter.android_flutter_host.FlutterEmbeddingActivity$Companion.init(FlutterEmbeddingActivity.kt:168) 
        at pro.truongsinh.flutter.android_flutter_host.MainActivity.onCreate(MainActivity.kt:17) 
        at android.app.Activity.performCreate(Activity.java:7802) 
        at android.app.Activity.performCreate(Activity.java:7791) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1299) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3243) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3407) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7343) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:933) 
Disconnected from the target VM, address: 'localhost:8606', transport: 'socket'
```

**Comment by truongsinh** [comment]
Figured out, with some caveats:

1. I have to use `lifecycle_version = "1.1.1"`, not `2.0.0`
```groovy
    def lifecycle_version = "1.1.1"
    implementation "android.arch.lifecycle:common-java8:$lifecycle_version"
```

2. Even with this, my example code still crashed (something like attempt to use destroyed engine), because I use shared cached Flutter enginer, and to fix this, have to override:
```kotlin
    override fun retainFlutterEngineAfterHostDestruction(): Boolean {
        return true
    }
```

Still caveats, it seems that the performance of transitioning between Kotlin activity and Flutter activity is much worse than before. Previously, with cached engine, it's really smooth. Even though low performance happens on debug mode only (not profile nor release), it might still be a performance degradation from previous version (1.5.4-hotfix)

**Comment by xster** [comment]
It's hard for us to make sure we follow up on comments on previously merged PRs. If there are still remaining issues or inconveniences with the flows, please file a new issue with repro steps.

**Comment by blasten** [comment]
@truongsinh Btw, we will start shipping Maven artifacts. As a result, these artifacts will become the source of truth for the engine dependencies such as `android.arch.lifecycle:common-java8`. Once this happens, there won't be more `NoClassDefFoundError`.

**Comment by truongsinh** [comment]
> @truongsinh Btw, we will start shipping Maven artifacts. As a result, these artifacts will become the source of truth for the engine dependencies such as `android.arch.lifecycle:common-java8`. Once this happens, there won't be more `NoClassDefFoundError`.

Do we have a github issue or PR to track this?

**Comment by blasten** [comment]
There's https://github.com/flutter/flutter/issues/11439

**Pull request merged**
